### DISCUSSION TITLE
Strange Behavior With CoreData and ForEachStore

### DISCUSSION THREAD

**Discussion by m-housh** [pointfreeco/swift-composable-architecture#2137] [Q&A]
First off, I'm not very confident in my core-data skills as I've never worked with it, but I have an upcoming project that I intend to use it for.  I built a toy todo application to play around with it, as well as experiment with CoreData inside a `swift-package`.

I thought that I had a working example using a plain old array and a `ForEach`, however now I'm not so sure as I was deleting all the items in between my experimentation's.

Anyway, my problem is that I start out with an empty array of `Todo`'s then load my Entity objects in from CoreData and convert them to a struct that stores the values along with the `NSManagedID`, my `State` get's populated, however my view does not reflect the state although it shows they were loaded using the `_printChanges()` on my reducer.

All of the online examples use the `@FetchRequest`, but I'm not sure if / how to use something like that in the TCA pattern,
so my dependency client object just loads the objects in and handles saving, updating, etc.

I'm not able to reproduce the problems using TCA without CoreData, so I'm ASSuming I'm doing something wrong with how I'm managing that, but at any rate I do feel it's strange behavior that the view does not change when the state changes when I'm observing the whole state.  I tried using a binding on my items, a view state object to observe changes, I've also tried getting back to what I thought was a working state just using the SwiftUI `ForEach` instead of a `ForEachStore`, but nothing seems to be working.

At any rate, [here](https://github.com/m-housh/TodosFramework)  is my toy Todo's app if there are any suggestions or if anyone has any projects they can share that use CoreData and TCA together, that would be awesome.  The `main` branch shows the problem when I first noticed it and uses a `ForEachStore` there's also a `for-each` branch that was me trying to get back to what I thought was a working state prior to my refactor of using a `ForEachStore`.

TIA.

**Comment by tgrapperon** [comment]
Hey @m-housh! 
I don't have the time to look at your example yet, but here are a few general recommendations: CoreData and TCA will be something complex to get working. Most of the issues come from the fact that `NSManagedObject` are:
- Reference types (classes), which mean that they don't change when you change one if their properties.
- Not thread-safe, and therefore should only be accessed on the `NSManagedObjectContext`'s thread that spawned them.

For this reason, CoreData will not work out of the box with TCA. One way to make it working while relatively preserving CoreData performance is to model each entity as a `Reducer/Domain`, with relations being standard TCA child domains relations, but this is not really an easy task.

In 2023, the only reason I find to use CoreData is to get automatic CloudKit synchronization across devices. If you don't need that and you're writing a TCA app, I'd advise using a persistence layer that relies on value types. You can check [GRDB](https://github.com/groue/GRDB.swift) for example.

**Reply by m-housh** [reply]
Thanks for the reply.  Yeah, I was wanting to use CloudKit synchronization in my next project which is why I'm playing with CoreData... I'm currently not using the managed object's themselves, I'm turning them into struct types that are used in my view layer then have a dependency that manages persisting them.

At any rate, the more I'm digging into the problem I don't think it's with CoreData, but something is going on where my State is changing or something.  The `printChanges` shows items are loaded, but it does not propagate to my state.  However if I allow the view to load first then  load items with a button, it works.

https://github.com/pointfreeco/swift-composable-architecture/assets/9305678/d6131400-2b6e-40e5-923a-74dac8be8887

**Comment by m-housh** [comment]
The more I'm digging into the problem I don't think it's with CoreData, but something is going on where my State is changing or something. The printChanges shows items are loaded, but it does not propagate to my state. However if I allow the view to load first then load items with a button, it works.

I added an ID to my state and when the view appears the ID is different than the ID after it has appeared and loaded the state, which seems like unexpected behavior, however I'm not sure if this is a SwiftUI problem or not.  Here's a video that shows the ID's printed to the console are different and a video that shows that the load button works once the view has appeared.

https://github.com/pointfreeco/swift-composable-architecture/assets/9305678/0a734bb8-6626-4f63-9e08-ae16d3dbed86


https://github.com/pointfreeco/swift-composable-architecture/assets/9305678/bc8878bf-873a-4e24-899e-213208a77c85

**Comment by jinwoong16** [comment]
hey @m-housh , I just found this post while looking for information about CoreData with TCA. 

I'm not sure if you've already solved that problem. The problem what I found was `@Environment(\.scenePhase) var scenePhase`.
I don't know why, but it was initializing the view and store in the window group of `TodosApp.swift` two or more times. 
As you mentioned, when the view is appeared, the app retrieved data from the CoreData correctly, but soon the state reinitialized to an empty state. 
Perhaps when the app enters the background, we need to find another way to save the data or figure out why this is happening.

By the way, a warning message about dissmissing the sheet view was appeared when I ran the app, I modified some codes like below:
```swift
// TodoListView.swift
...
case let .destination(.presented(.addTodo(.didSave(todo: todo)))):
  state.todos[id: todo.id] = .init(todo: todo)
  state.destination = nil // remove this
  return .send(.destination(.dismiss))
...

// AddTodoView.swift
...
@Dependency(\.dismiss) var dismiss // add this
...
case .saveButtonTapped:
  return .run { [todo = state.todo] send in
    ...
    await send(.didSave(todo: saved))
    await self.dismiss() // add this
  }
```

Lastly, if you find out the cause of this issue, please let me know, and thank you for sharing your code.
